### ISSUE TITLE
Interpolating data (again)

### ISSUE THREAD

**Issue by jtmbeta** [beOn/cili#12]
Hi Ben

Me again, requesting advice on data interpolation (just 4 days after you closed the last issue! sorry about that....) I've been playing around with cili's methods and comparing them to a common method used in the pupillometry literature, which is to use linear interpolation with a 100ms margin surrounding EyeLink blink events. With some blinks it makes no difference:

![test_interpolation1](https://cloud.githubusercontent.com/assets/11196943/20029880/68e2813e-a350-11e6-8e0b-f2d57822e3bd.png)

But with others, the value of the z-threshold method is obvious:

![test_interpolation2](https://cloud.githubusercontent.com/assets/11196943/20029882/7d0759fa-a350-11e6-9185-538ee4e4d34c.png)

I'm basically trying to justify using the z-threshold method in my thesis, and one thing that would be very helpful is if cili returned the adjusted indices for blink starts and ends when using the adjust_eyelink_recov_idxs=True option. Can you suggest the best approach for this?  

Also, do you think it would be handy to add another function to cili that goes something like:

interpolate_blinks_with_margin(samps, events, margin=100, interp_fields=['pup_l'])

I'm always a bit weary of changing the code when it already works so well, so I thought best to ask your advice on possible implementations first. 

All the best
Joel

**Comment by beOn** [comment]
I like this... though I think it would be better to add an optional "margins" named argument to the existing function, which will only be sued if specified.

**Comment by beOn** [comment]
Actually... let me give it a little more thought over the weekend.

**Comment by jtmbeta** [comment]
I've thought a bit more about this in the past few days. As you suggested in the previous issue, I interpolated data sets with and without the find_recovery option and it changes neither the shape nor significance of the data. In fact, there seems to be hardly any difference between these two methods and the fixed margin approach, so I guess the most important point is to have a sensible fixed rule for excluding participants on the basis of the amount of data that is interpolated because of blinks. After rummaging through the literature i've found recommendations for excluding participants who needed more than 15-18% of their trial data interpolated. To be able to get this, I needed an extra column in my ranges dataframe that indicated whether the sample had been altered by the interpolation routine, so I made the following adjustment to the mask

**Comment by jtmbeta** [comment]
I've thought a bit more about this in the past few days. As you suggested in the previous issue, I interpolated data sets with and without the find_recovery option and it changes neither the shape nor significance of the data. In fact, there seems to be hardly any difference between these two methods and the fixed margin approach, so I guess the most important point is to have a sensible fixed rule for excluding participants on the basis of the amount of data that is interpolated because of blinks. After rummaging through the literature i've found recommendations for excluding participants who needed more than 15-18% of their trial data interpolated. To be able to get this, I needed an extra column in my ranges dataframe that indicated whether the sample had been altered by the interpolation routine, so I made the following adjustment to the mask_eyelink_blinks function (last two lines before return):

`def mask_eyelink_blinks(samples, events, mask_fields=["pup_l"], find_recovery=True):

```
samps = samples.copy(deep=True)
indices = get_eyelink_mask_idxs(samps, events, find_recovery=find_recovery)
samps.loc[indices, mask_fields] = float('nan')
samps['interpolated'] = 0
samps.loc[indices, 'interpolated'] = 1
return samps`
```

This means that after using extract_events, I can print a quick ranges.interpolated.value_counts(normalize=True) to decide whether the participant should be rejected in accordance with the exclusion rule (whatever that is in the end). Also, the extra info makes it easy to reject individual trials on the basis of blinks and interpolated data, overall, or at specific time points. This alteration meant I also had to change interp_zeros slightly because of the way it sets all columns and not just the 'mask_fields' to 'nan' (is it meant to do that?). 

So to summarize, I find that having that extra info in the ranges dataframe is useful, but i'm not sure if my alteration was the best way to do this, or if there is a more elegant solution. Also, I think the 100ms margin option would be nice option to have, because this is the approach in quite a few publications. 

Finally, please don't waste your weekend on this! I've already benefited much from this package, so if you can give a few comments, maybe I could submit a pull request? It would be my first...

**Comment by beOn** [comment]
The extra 'interpolated' field sounds useful. Please submit a PR.

That's not how interp_zeros was meant to work... but I think I see where it's messed up. Looks to me like the problem is in cleanup.py lines 132-133:

```python
for f in mask_fields:
        samps[samps[f] == 0] = float("nan")
```

Indixing issue, non? If you have a fix for this ready, please include it in the PR.

(Side note: If only test_cleanup.py were a thing. Someone (me) should get on that.)
